crypto/bls_test.go

```diff
@@ -299,3 +299,171 @@ func TestRemovePubKeys(t *testing.T) {
+	seed := make([]byte, KeyGenSeedMinLenBLSBLS12381)
+	expectedValid := make([]bool, 0, sigsNum)
+
+	// create the signatures
+	for i := 0; i < sigsNum; i++ {
+		sk := randomSK(t, seed)
+		s, err := sk.Sign(input, kmac)
+		require.NoError(t, err)
+		sigs = append(sigs, s)
+		sks = append(sks, sk)
+		pks = append(pks, sk.PublicKey())
+		expectedValid = append(expectedValid, true)
+	}
+
+	// Batch verify the signatures
+	// all signatures are valid
+	valid, err := BatchVerifySignaturesOneMessage(pks, sigs, input, kmac)
+	require.NoError(t, err)
+	assert.Equal(t, valid, expectedValid,
+		fmt.Sprintf("Verification of %s failed, private keys are %s, input is %x, results is %v",
+			sigs, sks, input, valid))
+
+	// some signatures are invalid
+	invalidSigsNum := mrand.Intn(sigsNum-1) + 1 // pick a random number of invalid signatures
+	indices := make([]int, 0, sigsNum)          // pick invalidSigsNum random indices
+	for i := 0; i < sigsNum; i++ {
+		indices = append(indices, i)
+	}
+	mrand.Shuffle(sigsNum, func(i, j int) {
+		indices[i], indices[j] = indices[j], indices[i]
+	})
+
+	for i := 0; i < invalidSigsNum; i++ { // alter invalidSigsNum random signatures
+		alterSignature(sigs[indices[i]])
+		expectedValid[indices[i]] = false
+	}
+
+	valid, err = BatchVerifySignaturesOneMessage(pks, sigs, input, kmac)
+	require.NoError(t, err)
+	assert.Equal(t, expectedValid, valid,
+		fmt.Sprintf("Verification of %s failed\n private keys are %s\n input is %x\n results is %v",
+			sigs, sks, input, valid))
+
+	// all signatures are invalid
+	for i := invalidSigsNum; i < sigsNum; i++ { // alter the remaining random signatures
+		alterSignature(sigs[indices[i]])
+		expectedValid[indices[i]] = false
+		if i%5 == 0 {
+			sigs[indices[i]] = sigs[indices[i]][:3] // test the short signatures
+		}
+	}
+
+	valid, err = BatchVerifySignaturesOneMessage(pks, sigs, input, kmac)
+	require.NoError(t, err)
+	assert.Equal(t, valid, expectedValid,
+		fmt.Sprintf("Verification of %s failed, private keys are %s, input is %x, results is %v",
+			sigs, sks, input, valid))
+
+	// test the empty list case
+	valid, err = BatchVerifySignaturesOneMessage(pks[:0], sigs[:0], input, kmac)
+	require.Error(t, err)
+	assert.Equal(t, valid, []bool{},
+		fmt.Sprintf("verification should fail with empty list key, got %v", valid))
+	// test incorrect inputs
+	valid, err = BatchVerifySignaturesOneMessage(pks[:len(pks)-1], sigs, input, kmac)
+	require.Error(t, err)
+	assert.Equal(t, valid, []bool{},
+		fmt.Sprintf("verification should fail with incorrect input lenghts, got %v", valid))
+	// test wrong hasher
+	for i := 0; i < sigsNum; i++ {
+		expectedValid[i] = false
+	}
+	valid, err = BatchVerifySignaturesOneMessage(pks, sigs, input, nil)
+	require.Error(t, err)
+	//require.Nil(t, aggSig)
+	assert.Equal(t, valid, expectedValid,
+		fmt.Sprintf("verification should fail with incorrect input lenghts, got %v", valid))
+}
+
+// alter or fix a signature
+func alterSignature(s Signature) {
```

Comment:
Much of the test assumes this alters the signature, I'm assuming this function works as intended

Reply:
The function mainly switches a bit in the signature, and that causes the signature to be invalid for a different reason each time depending on the original value, which allows testing different failing scenarios.